policy/aggregation_type_compress.go

```diff
@@ -0,0 +1,147 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package policy
+
+import (
+	"fmt"
+
+	"github.com/willf/bitset"
+)
+
+type validateFn func(aggType AggregationType) bool
+type errorFn func(aggType AggregationType) error
+
+var (
+	validateGaugeFn   = validateFn(func(aggType AggregationType) bool { return aggType.IsValidForGauge() })
+	validateCounterFn = validateFn(func(aggType AggregationType) bool { return aggType.IsValidForCounter() })
+	validateTimerFn   = validateFn(func(aggType AggregationType) bool { return aggType.IsValidForTimer() })
+	validateAllFn     = validateFn(func(aggType AggregationType) bool { return aggType.IsValid() })
+
+	invalidGaugeTypeErrorFn = func(aggType AggregationType) error {
+		return fmt.Errorf("invalid AggregationType: %s for Gauge", aggType.String())
+	}
+	invalidCounterTypeErrorFn = func(aggType AggregationType) error {
+		return fmt.Errorf("invalid AggregationType: %s for Counter", aggType.String())
+	}
+	invalidTimerTypeErrorFn = func(aggType AggregationType) error {
+		return fmt.Errorf("invalid AggregationType: %s for Timer", aggType.String())
+	}
+	invalidTypeErrorFn = func(aggType AggregationType) error {
+		return fmt.Errorf("invalid AggregationType: %s", aggType.String())
+	}
+)
+
+// AggregationTypeCompressor can compress AggregationTypes.
+type AggregationTypeCompressor interface {
+	Compress(aggTypes AggregationTypes) (CompressedAggregationTypes, error)
+}
+
+// AggregationTypeDecompressor can decompress CompressedAggregationTypes for different types of metrics.
+type AggregationTypeDecompressor interface {
+	// Decompress decompresses aggregation types, return error if any invalid aggregation type is found.
+	Decompress(compressed CompressedAggregationTypes) (AggregationTypes, error)
+	// DecompressForCounter decompresses aggregation types,
```

Comment:
Yeah doing the type specific decompress allows us to fail earlier in the decompression, but it does make the interface less clean, removed them